src/wbase/ChannelShared.cc

```diff
@@ -0,0 +1,337 @@
+
+using namespace std;
+
+namespace {
+LOG_LOGGER _log = LOG_GET("lsst.qserv.wbase.ChannelShared");
+}
+
+namespace lsst::qserv::wbase {
+
+atomic<uint64_t> ChannelShared::scsSeqId{0};
+
+ChannelShared::ChannelShared(shared_ptr<wbase::SendChannel> const& sendChannel,
+                             shared_ptr<wcontrol::TransmitMgr> const& transmitMgr, qmeta::CzarId czarId)
+        : _sendChannel(sendChannel), _transmitMgr(transmitMgr), _czarId(czarId), _scsId(scsSeqId++) {
+    if (_sendChannel == nullptr) {
+        throw util::Bug(ERR_LOC, "ChannelShared constructor given nullptr");
+    }
+}
+
+ChannelShared::~ChannelShared() {
+    if (_sendChannel != nullptr) {
+        _sendChannel->setDestroying();
+        if (!_sendChannel->isDead()) {
+            _sendChannel->kill("~ChannelShared()");
+        }
+    }
+}
+
+bool ChannelShared::send(char const* buf, int bufLen) {
+    lock_guard<mutex> const streamMutexLock(_streamMutex);
+    return _sendChannel->send(buf, bufLen);
+}
+
+bool ChannelShared::sendError(string const& msg, int code) {
+    lock_guard<mutex> const streamMutexLock(_streamMutex);
+    return _sendChannel->sendError(msg, code);
+}
+
+bool ChannelShared::sendFile(int fd, wbase::SendChannel::Size fSize) {
+    lock_guard<mutex> const streamMutexLock(_streamMutex);
+    return _sendChannel->sendFile(fd, fSize);
+}
+
+bool ChannelShared::sendStream(xrdsvc::StreamBuffer::Ptr const& sBuf, bool last, int scsSeq) {
+    lock_guard<mutex> const streamMutexLock(_streamMutex);
+    return _sendChannel->sendStream(sBuf, last, scsSeq);
+}
+
+bool ChannelShared::kill(string const& note) {
+    lock_guard<mutex> const streamMutexLock(_streamMutex);
+    return _kill(streamMutexLock, note);
+}
+
+bool ChannelShared::isDead() {
+    if (_sendChannel == nullptr) return true;
+    return _sendChannel->isDead();
+}
+
+void ChannelShared::setTaskCount(int taskCount) { _taskCount = taskCount; }
+
+bool ChannelShared::transmitTaskLast() {
+    lock_guard<mutex> const streamMutexLock(_streamMutex);
+    ++_lastCount;
+    bool lastTaskDone = _lastCount >= _taskCount;
+    return lastTaskDone;
+}
+
+bool ChannelShared::_kill(lock_guard<mutex> const& streamMutexLock, string const& note) {
+    LOGS(_log, LOG_LVL_DEBUG, "ChannelShared::kill() called " << note);
+    bool ret = _sendChannel->kill(note);
+    _lastRecvd = true;
+    return ret;
+}
+
+string ChannelShared::makeIdStr(int qId, int jId) {
+    string str("QID" + (qId == 0 ? "" : to_string(qId) + "#" + to_string(jId)));
+    return str;
+}
+
+uint64_t ChannelShared::getSeq() const { return _sendChannel->getSeq(); }
+
+string ChannelShared::dumpTransmit() const {
+    lock_guard<mutex> const tMtxLock(tMtx);
+    return dumpTransmit(tMtxLock);
+}
+
+bool ChannelShared::buildAndTransmitError(util::MultiError& multiErr, Task::Ptr const& task, bool cancelled) {
+    auto qId = task->getQueryId();
+    bool scanInteractive = true;
+    waitTransmitLock(scanInteractive, qId);
```

Comment:
I will see what I can do to get this implemented.

Reply:
I have evaluated the code of `waitTransmitLock`. I propose eliminating this method in a separate "code cleanup" JIRA ticket after merging the current PR and using the new code in production. After that, we would be able to eliminate the SSI-based result delivery path. This will eliminate Any need in having the "transmit manager" which was introduced as a flow-control mechanism to limit the memory usage & XROOTD/SSI load.

If we do this step now for the `FileChannelShared` then we will end up with a more complicated code. As you mentioned in your comment, calling this method from `FileChannelShared` won't do any damage.